File: packages/sentinel-py/src/ameva_sentinel/core/policy_registry.py
"""
In-Memory Cost Policy Registry for Python.
"""

from typing import Dict, Optional, Any
from .budget_types import CostPolicyConfig, RouteCostPolicy, HttpMethod
from .canonical import (
    SAFE_FALLBACK_COST_POLICY,
    compute_policy_checksum,
    validate_cost_policy,
)


class CostPolicyRegistry:
    def __init__(self, raw_policy: Optional[Any] = None):
        if raw_policy is None:
            self.config = SAFE_FALLBACK_COST_POLICY
            identity = compute_policy_checksum(self.config)
            self.checksum = identity.checksum_sha256
            self.display_checksum = identity.display_checksum
        else:
            try:
                self.config = validate_cost_policy(raw_policy)
                identity = compute_policy_checksum(self.config)
                self.checksum = identity.checksum_sha256
                self.display_checksum = identity.display_checksum
            except Exception:
                self.config = SAFE_FALLBACK_COST_POLICY
                identity = compute_policy_checksum(self.config)
                self.checksum = identity.checksum_sha256
                self.display_checksum = identity.display_checksum

        self._exact_routes: Dict[str, RouteCostPolicy] = {}
        self._wildcard_routes: Dict[str, RouteCostPolicy] = {}
        self._index_routes()

    def _index_routes(self) -> None:
        self._exact_routes.clear()
        self._wildcard_routes.clear()

        for r in self.config.routes:
            clean_path = r.path.rstrip("/") or "/"
            clean_method = r.method.upper()

            if clean_method == "*":
                self._wildcard_routes[clean_path] = r
            else:
                self._exact_routes[f"{clean_method}:{clean_path}"] = r

    def get_route_policy(self, method: str, path: str) -> RouteCostPolicy:
        clean_path = path.split("?")[0].rstrip("/") or "/"
        clean_method = method.upper()

        exact_key = f"{clean_method}:{clean_path}"
        if exact_key in self._exact_routes:
            return self._exact_routes[exact_key]

        if clean_path in self._wildcard_routes:
            return self._wildcard_routes[clean_path]

        d = self.config.defaults
        return RouteCostPolicy(
            method=clean_method,  # type: ignore
            path=clean_path,
            cost=d.cost,
            authentication=d.authentication,
            page_size_default=d.page_size_default,
            page_size_max=d.page_size_max,
            query_timeout_ms=d.query_timeout_ms,
            lock_timeout_ms=d.lock_timeout_ms,
            max_data_points=d.max_data_points,
            max_request_body_bytes=d.max_request_body_bytes,
            max_response_bytes=d.max_response_bytes,
            failure_mode=d.failure_mode,
            shadow_mode=d.shadow_mode,
        )

File: packages/risk-core/src/policy-registry.ts
/**
 * @file policy-registry.ts
 * In-Memory Cost Policy Registry for deterministic route matching and checksum resolution.
 */

import { CostPolicyConfig, RouteCostPolicy, HttpMethod } from './budget-types.js';
import {
  SAFE_FALLBACK_COST_POLICY,
  computePolicyChecksum,
  validateCostPolicy
} from './policy-canonical.js';

export class CostPolicyRegistry {
  public readonly config: CostPolicyConfig;
  public readonly checksum: string; // full 64-char SHA-256
  public readonly displayChecksum: string; // 16-char display
  private readonly exactRoutes = new Map<string, RouteCostPolicy>();
  private readonly wildcardRoutes = new Map<string, RouteCostPolicy>();

  constructor(rawPolicy?: unknown) {
    if (!rawPolicy) {
      this.config = SAFE_FALLBACK_COST_POLICY;
      const identity = computePolicyChecksum(this.config);
      this.checksum = identity.checksumSha256;
      this.displayChecksum = identity.displayChecksum;
    } else {
      try {
        this.config = validateCostPolicy(rawPolicy);
        const identity = computePolicyChecksum(this.config);
        this.checksum = identity.checksumSha256;
        this.displayChecksum = identity.displayChecksum;
      } catch (err: unknown) {
        // [AUDIT FIX] Silent fallback removed. Operators must be notified when
        // a supplied policy is rejected so they can diagnose deployment issues.
        const reason = err instanceof Error ? err.message : String(err);
        console.error(
          `[ameva-sentinel] CostPolicyRegistry: supplied policy failed validation and was rejected. ` +
          `Falling back to SAFE_FALLBACK_COST_POLICY. ` +
          `Reason: ${reason}`
        );
        this.config = SAFE_FALLBACK_COST_POLICY;
        const identity = computePolicyChecksum(this.config);
        this.checksum = identity.checksumSha256;
        this.displayChecksum = identity.displayChecksum;
      }
    }

    this.indexRoutes();
  }

  private indexRoutes(): void {
    this.exactRoutes.clear();
    this.wildcardRoutes.clear();

    for (const r of this.config.routes || []) {
      const cleanPath = r.path.replace(/\/+$/, '') || '/';
      const cleanMethod = r.method.toUpperCase();

      if (cleanMethod === '*') {
        this.wildcardRoutes.set(cleanPath, r);
      } else {
        this.exactRoutes.set(`${cleanMethod}:${cleanPath}`, r);
      }
    }
  }

  public getRoutePolicy(method: string, path: string): RouteCostPolicy {
    const cleanPath = path.split('?')[0].replace(/\/+$/, '') || '/';
    const cleanMethod = method.toUpperCase() as HttpMethod;

    const exact = this.exactRoutes.get(`${cleanMethod}:${cleanPath}`);
    if (exact) return exact;

    const wildcard = this.wildcardRoutes.get(cleanPath);
    if (wildcard) return wildcard;

    const d = this.config.defaults;
    return {
      method: cleanMethod,
      path: cleanPath,
      cost: d.cost,
      authentication: d.authentication,
      page_size_default: d.page_size_default,
      page_size_max: d.page_size_max,
      query_timeout_ms: d.query_timeout_ms,
      lock_timeout_ms: d.lock_timeout_ms,
      max_data_points: d.max_data_points,
      max_request_body_bytes: d.max_request_body_bytes,
      max_response_bytes: d.max_response_bytes,
      failure_mode: d.failure_mode,
      shadow_mode: d.shadow_mode
    };
  }
}
